File: libraries/OclProcess.go
package OclProcess

// import "container/list"
// import "fmt"
import "../ocl"
import "os"
// import "math"
import "os/exec"
import "sync"
import "time"


type OclProcess struct {
  name string 
  priority int 
  executes interface{}
  command *exec.Cmd

  deadline int64 
  delay int64 
  period int64 
}

func createOclProcess() *OclProcess {
  var res *OclProcess
  res = &OclProcess{}
  return res
}

func NewOclProcess(obj interface{}, s string) *OclProcess {
  var result *OclProcess
  result = createOclProcess()
  result.executes = obj
  result.name = s
  return result
}

func (self *OclProcess) SetName(s string) { 
  self.name = s
}

func (self *OclProcess) GetName() string { 
  return self.name
}

func GetRuntime() *OclProcess {
  result := NewOclProcess(nil, "System")
  return result
}

func Notify(c *sync.Cond) {
  c.Signal()
}

func NotifyAll(c *sync.Cond) {
  c.Broadcast()
}

func Wait(c *sync.Cond, t float64) {
  c.Wait()
}


func (self *OclProcess) Start() {
  if self.executes == nil { 
    cmd := exec.Command(self.name)
    self.command = cmd
    cmd.Start()
  } else { 
    go ((self.executes).(func() int))()
  }
} 

func (self *OclProcess) Run() {
  if self.executes == nil { 
    cmd := exec.Command(self.name)
    self.command = cmd
    cmd.Run()
  } else { 
    go ((self.executes).(func() int))()
  }
} 

func (self *OclProcess) WaitFor() int {
  if self.command != nil { 
    err := self.command.Wait()
    if err != nil { 
      return -1
    } 
    return 0
  } 
  return 0
}

func (self *OclProcess) SetDeadline(d int64) {
  self.deadline = d
} 

func (self *OclProcess) SetDelay(d int64) {
  self.delay = d
}

func (self *OclProcess) SetPeriod(p int64) {
  self.period = p
}

func (self *OclProcess) GetDeadline() int64 {
  return self.deadline
} 

func (self *OclProcess) GetDelay() int64 {
  return self.delay
}

func (self *OclProcess) GetPeriod() int64 {
  return self.period
}

    
func GetEnvironmentProperty(vbl string) string {
  return os.Getenv(vbl)
} 

func GetEnvironmentProperties() map[string]string { 
  res := make(map[string]string)
  vars := os.Environ()
  for i := 0; i < len(vars); i++ {
      strs := ocl.Split(vars[i],"=")
      if strs.Len() > 1 { 
        res[ocl.First(strs).(string)] = ocl.Last(strs).(string)
      } 
  } 
  return res
}

func Sleep(n int64) { 
  time.Sleep(time.Duration(n)*time.Millisecond)
}

func Exit(n int) { 
  os.Exit(n)
} 

File: libraries/OclProcess.swift
import Foundation
import Darwin


protocol Runnable
{ func run() -> Void } 


class OclProcess
{ var name : String = ""
  var priority : Int = 5 
  var process : Thread? = nil 
  var osprocess : Process? = nil
  private static var instance : OclProcess?

  static func defaultInstance() -> OclProcess
  { if instance == nil
    { instance = OclProcess() } 
    return instance!
  } 

  static func getRuntime() -> OclProcess
  { return OclProcess.defaultInstance() } 

  static func notify(obj : NSCondition) 
  { obj.signal() }

  static func notifyAll(obj : NSCondition) 
  { obj.broadcast() } 

  static func wait(obj : NSCondition, t : Double) 
  { if t <= 0
    { obj.wait() }
    else 
    { let d = Date(timeIntervalSinceNow: t/1000.0)
      obj.wait(until: d)
    }
  } 

  func waitFor() -> Int
  { if osprocess != nil
    { osprocess!.waitUntilExit() } 
    return 0
  } 

  /* BlockOperation could be used here
     with operation!.waitUntilFinished() */ 

  static func newOclProcess(obj : Runnable, s : String) -> OclProcess
  { let p = OclProcess()
    p.name = s
    p.priority = 5
    let actualp = 
      Thread(block: { obj.run() })
    p.process = actualp
    return p
  } 

  static func newOclProcess(obj : Any?, s : String) -> OclProcess
  { let p = OclProcess()
    p.name = s
    p.priority = 5
    let actualp = Process()
    actualp.executableURL = URL(fileURLWithPath: s)
    p.osprocess = actualp
    return p
  } 

  static func activeCount() -> Int
  { return 1 } 

  static func currentThread() -> OclProcess
  { let res = OclProcess()
    res.name = "Current thread"
    res.process = Thread.current
    return res
  } 
  
  static func allActiveThreads() -> [OclProcess]
  { return [] } 

  static func sleep(n : Int64)
  { Thread.sleep(forTimeInterval: Double(n)/1000.0) }  

  func getName() -> String
  { return name } 

  func setName(nme : String)
  { name = nme }

  func getPriority() -> Int
  { return priority } 

  func run() 
  { if process != nil
    { process!.start() }
    else if osprocess != nil
    { try? osprocess!.run() } 
  } 

  func start()
  { if process != nil
    { process!.start() }
    else if osprocess != nil
    { try? osprocess!.run() } 
  }

  func isAlive() -> Bool
  { if process != nil 
    { return process!.isExecuting } 
    if osprocess != nil
    { return osprocess!.isRunning } 
    return false
  } 

  func isDaemon() -> Bool
  { return false }

  func join(ms : Double) 
  { if osprocess != nil
    { osprocess!.waitUntilExit() }
  } 

  func interrupt() 
  { } 

  func destroy() 
  { if process != nil
    { process!.cancel() }
    if osprocess != nil
    { osprocess!.terminate() } 
  } 

  static func getEnvironmentProperty(vbl : String) -> String
  { let p = ProcessInfo.processInfo
    return p.environment[vbl] ?? ""
  } 
  
  static func getEnvironmentProperties() -> [String:String]
  { let p = ProcessInfo.processInfo
    return p.environment
  } 
  
  static func exit(n : Int)
  { Thread.exit() } 
}

func displayOclProcess(_ s: OclProcess)
{ print(String(describing: s)) } 

